macaw-aarch32/src/Data/Macaw/ARM/Arch.hs

```diff
@@ -698,8 +699,9 @@ t32InstructionMatcher (ThumbDis.Instruction opc operands) =
                                    <*> G.getRegVal (ARMReg.ARMGlobalBV (ASL.knownGlobalRef @"_R6"))
                                    <*> G.getRegVal (ARMReg.ARMGlobalBV (ASL.knownGlobalRef @"_R7"))
             res <- G.addExpr =<< evalArchFn sc
-            G.setRegVal r0 =<< G.addExpr (G.AppExpr (MC.TupleField PC.knownRepr res PL.index0))
-            G.setRegVal r1 =<< G.addExpr (G.AppExpr (MC.TupleField PC.knownRepr res PL.index1))
+            -- res is a tuple of form (R1, R0)
+            G.setRegVal r1 =<< G.addExpr (G.AppExpr (MC.TupleField PC.knownRepr res PL.index0))
+            G.setRegVal r0 =<< G.addExpr (G.AppExpr (MC.TupleField PC.knownRepr res PL.index1))
```

Comment:
Ah, now I can see why this was so easy to overlook—it's because we're using `index0` and `index1` in what seems like an intuitive order, but the result doesn't match that intuition. Perhaps we should explicitly match on `res` rather than using indices to make this more obvious, e.g.,

```hs
let r1 :> r0 :> Nil = res
G.setRegVal r1
G.setRegVal r2
```

It might also be worth writing a Note to explain why we end up with seemingly the opposite order you'd expect for returned values, explaining the impedance mismatch between `Data.Parameterized.List` and `Assignment`.

Reply:
Agreed

Reply:
The pattern match doesn't work because `res` is an `Expr ids ('TupleType '[BVType 64, BVType 64])` and not just a `Data.Parameterized.List`.  Getting to the individual list would require evaluating the `Expr`, converting the `Value` into a `PL.List` (this looks potentially gross), matching on the list, then building up new `Expr`s for each field.  I think this could be more confusing than the current approach as the match on the list will be hidden among a bunch of evaluation/extraction/wrapping logic.

I will definitely add a comment about the impedance mismatch though.